File: leetcode/103-BinaryTreeZigzagLevelOrderTraversal/binaryTreeZigzagLevelOrderTraversal.py
 # [Source] 
 #          
 # https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/description/
 #          
 # [Problem Description]
 #          
 # Given a binary tree, return the zigzag level order traversal of its nodes'
 # values. (ie, from left to right, then right to left for the next level and
 # alternate between).
 # 
 # For example:
 # Given binary tree [3,9,20,null,null,15,7],
 # 
 # 3
 # / \
 # 9  20
 # /  \
 # 15   7
 # 
 # return its zigzag level order traversal as:
 # 
 # [
 # [3],
 # [20,9],
 # [15,7]
 # ]
 #          
 # [Comments]
 #          
 #          
 #          
 # [Companies]
 # Bloomberg


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        # Reference: http://algorithms.tutorialhorizon.com/level-order-traversal-in-zig-zag-pattern-or-print-in-spiral-pattern/
        evenLevel = False # Root level is counted as level 0
        queue = []
        res = []
        if not root:
            return []
        queue.append(root)
        while queue:
            levelNodes = len(queue)
            al = []
            while levelNodes > 0:
                n = queue.pop(0)
                al.append(n.val)
                if (n.left):
                    queue.append(n.left)
                if (n.right):
                    queue.append(n.right)
                levelNodes -= 1
            if evenLevel:
                al.reverse()
                res.append(al)
                evenLevel = False
            else:
                res.append(al)
                evenLevel = True
        return res

File: leetcode/103-BinaryTreeZigzagLevelOrderTraversal/binaryTreeZigzagLevelOrderTraversal.cc
/*************************************************************************
  > File Name:       binaryTreeZigzagLevelOrderTraversal.cc
  > Author:          Zeyuan Hu
  > Mail:            iamzeyuanhu@utexas.edu
  > Created Time:    10/24/18
  > Description:
    
    C++ Solution
 ************************************************************************/


/**
* Definition for a binary tree node.
* struct TreeNode {
*     int val;
*     TreeNode *left;
*     TreeNode *right;
*     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
* };
*/

#include "cppinclude/bt.h"
#include <vector>
#include <queue>

using namespace std;

class Solution
{
public:
    vector<vector<int>>
    zigzagLevelOrder(TreeNode *root)
    {
        queue<TreeNode *> q;
        vector<vector<int>> res;
        // level counting starts with 0 (i.e. root in level 0)
        bool evenLevel = true;
        if (root == nullptr)
        {
            return {};
        }
        q.push(root);
        while (!q.empty())
        {
            int len = q.size();
            vector<int> tmp;
            for (int i = 0; i < len; ++i)
            {
                auto node = q.front();
                q.pop();
                tmp.push_back(node->val);
                if (node->left)
                {
                    q.push(node->left);
                }
                if (node->right)
                {
                    q.push(node->right);
                }
            }
            if (evenLevel)
            {
                res.push_back(tmp);
                evenLevel = false;
            }
            else
            {
                reverse(tmp.begin(), tmp.end());
                res.push_back(tmp);
                evenLevel = true;
            }
        }
        return res;
    }
};

using ptr2zigzagLevelOrder = vector<vector<int>> (Solution::*)(TreeNode *);

void
test(ptr2zigzagLevelOrder pfcn)
{
    Solution sol;
    BT bt;
    auto root = bt.list2Tree(vector<int>() = {3, 9, 20, NULLPTR, NULLPTR, 15, 7});
    (sol.*pfcn)(root);
    // TODO: Add assertions
}

int
main()
{
    ptr2zigzagLevelOrder pfcn = &Solution::zigzagLevelOrder;
    test(pfcn);
}
